docs: misc content & guide updates (#2359)
iterative/dvc.org: content/docs/start/data-and-model-versioning.md

Git.'
---

# Data Versioning

How cool would it be to make Git handle arbitrary large files and directories
with the same performance as with small code files? Imagine doing a `git clone`
and seeing data files and machine learning models in the workspace. Or switching
to a different version of a 100Gb file in less than a second with a
`git checkout`.

> The foundation of DVC consists of a few commands which you can run along with

Reviewer: Restrictive clause (keep "that"). A good rule of thumb I use: if it should have a comma before the clause, then use "which" (prob not the case here).

> The foundation of DVC consists of a few commands which you can run along with

Author: again; bordering on unrestrictive -> which (and defaulting to restrictive -> personal style choice).

I used "which" to signify immediate explication - i.e. "that" could refer to "foundation" or "commands," but "which" can be thought of as a shorthand for "which in turn" which can only refer to "commands."

> The foundation of DVC consists of a few commands which you can run along with

Reviewer: I tried reading the original sentence a few times and can't perceive "that" referring to "foundation". Wouldn't "which" have the same problem though?

Anyway. Since we're leaving the other ones you changed let's also resolve here for consistency. 👍

to a different version of a 100Gb file in less than a second with a
`git checkout`.

> The foundation of DVC consists of a few commands which you can run along with

Reviewer: What about just
```suggestion
The foundation of DVC consists of a few commands which can run along with
```
I'm just getting a feeling of "which" overuse. We can ask others to proof read if needed.

Reviewer: Gonna commit this one to try and move along, feel free to go back to this.

Reviewer: > Committed, please `git pull`.

### ⚙️ Expand to get an example dataset.

> Having initialized a project in the previous section, get the data file which we

Reviewer: Hmmm. Also not sure about "which" here ("which we will be using" isn't super readable). But the current text isn't great either. Maybe rewrite the whole paragraph?

Author: reworded

```

DVC stores information about the added file (or a directory) in a special `.dvc`

> file named `data/data.xml.dvc` - a small text file with a human-readable

Reviewer: ```suggestion
file named `data/data.xml.dvc` — a small text file with a human-readable
```

Author: lol corrected

Reviewer: Haha thanks!

[format](/doc/user-guide/project-structure/dvc-files). This metadata file can be
easily versioned like source code with Git. The original data, meanwhile, is

> listed in `.gitignore`:

Reviewer: Good rephrase, but "as a **placeholder** for the original data" was a key idea.

Reviewer: Also ``is listed in `.gitignore`:`` looks like we are going to demo the .gitignore contents below (which we don't do).

Author: reworded

### 💡 Expand to see what happens under the hood.

`dvc add` moved the data to the project's <abbr>cache</abbr>, and

> <abbr>linked</abbr> it back to the <abbr>workspace</abbr>.

Reviewer: I'd leave this for the basic concepts PR. I'm not sure file linking is a basic concepts. Maybe

Reviewer: I guess it can be! And I like this change. But it's not a copy edit so it prob belongs to a separate PR.

    path: data.xml
```

</details>

## Storing and sharing

You can upload DVC-tracked data or model files with `dvc push`, so they're
safely stored [remotely](/doc/command-reference/remote). This also means they
can be retrieved on other environments later with `dvc pull`. First, we need to

> setup a storage provider:

Reviewer: Agree that "a storage" sounds incomplete. How about a "storage location" or a "[DVC remote](https://dvc.org/doc/command-reference/remote)" for terminology consistency? We use those terms a lot (not really "provider").

Author: changed

```dvc
$ dvc remote add -d storage s3://mybucket/dvcstore
$ git add .dvc/config
$ git commit -m "Configure remote storage"
```

> DVC supports many remote storage types, including: Google Drive, Amazon S3,

> > Azure Blob Storage, Google Cloud Storage, Aliyun OSS, SSH, HDFS, and HTTP. See

Reviewer: 👍 but I'd either remove a few examples or keep the previous form and make it inclusive by adding the one main missing type (WebDAV).

Author: Would be hard to keep up-to-date. I prefer adding everything current whenever revised but keep the language leaving room for future additions. What would you suggest dropping from the examples? Aliyun? One of the Googles? HDFS?

Reviewer: Remotes don't change that often and we have to maintain other full lists anyway. But fine.

I'd say keep S3, SSH, GDrive, maybe GCS, Azure, HDFS (in that order). I have the impression that those are the most use ones.

> `dvc remote add` for more details and examples.

<details>

> ### ⚙️ Expand to set up a remote storage provider ☁

Reviewer: I think "setup remote storage" though it doesn't sounds as nice is better since it's the term we have everywhere. Feel free to make an issue to discuss, decide, and change the terminology throughout though but the GS shouold prob be very consistent with existing terminology and "remote storage" is the most common form atm. Also, always happy to add some elegant emojis like here, but unfort remote storage isn't always on the cloud so it may be misleading.

Reviewer: > "remote storage" is the most common form atm.

Also "DVC remote" and "data remote" (the last one at least should def. be revised IMO).

Author: changed

    path: data.xml
```

</details>

## Storing and sharing

You can upload DVC-tracked data or model files with `dvc push`, so they're
safely stored [remotely](/doc/command-reference/remote). This also means they
can be retrieved on other environments later with `dvc pull`. First, we need to

> setup a storage location:

Reviewer: The term "location" works here with as it's somewhat explanatory along with the first mention of "remote storage":
```suggestion
setup a remote storage location:
```

```dvc
$ dvc remote add -d storage s3://mybucket/dvcstore
$ git add .dvc/config
$ git commit -m "Configure remote storage"
```

> DVC supports many remote storage types, including: Google Drive, Amazon S3,
> Azure Blob Storage, Google Cloud Storage, Aliyun OSS, SSH, HDFS, and HTTP. See
> `dvc remote add` for more details and examples.

<details>

> ### ⚙️ Expand to set up a remote storage location.

Reviewer: But I think down here we can just remove the article and avoid "location":
```suggestion
### ⚙️ Expand to set up remote storage.
```

<details>

> ### ⚙️ Expand to refresh the project ⟳

Reviewer: I think they meant "dynamite".
"Refresh" sounds like re-loading the same data to me (wouldn't remove anything).

If we can't find the right phrase, maybe be explicit like "delete cached data".

Author: what about "delete locally cached data?"

Reviewer: Actually, checking https://dvc.org/doc/start/data-and-model-versioning I don't think we should use the term "cache" here as it's not visible anywhere else, probably intentionally. Maybe keep the original intention for now ("blow up") or go for something like "clear the storage".

<details>

> ### ⚙️ Expand to delete locally cached data.

Reviewer: > From https://github.com/iterative/dvc.org/pull/2359#discussion_r612877928:

I don't think we should use the term "cache" here as it's not visible anywhere else, probably intentionally. Maybe keep the original intention for now ("blow up") or go for something like "clear the storage".

Author: Er... the word "cache" is used a lot on the same page.

```

> 📖 See also

> > [sharing data and model files](/doc/use-cases/sharing-data-and-model-files)

Reviewer: Same comment about keeping title capitalization.

## Switching between versions

> The regular workflow is to use `git checkout` first (to switch a branch or
> checkout a commit/revision of a `.dvc` file) and then run `dvc checkout` to sync

Reviewer: Prob better, yes 👍 Though the parenthesis is still kind of long. Consider completely rewriting. For now:
```suggestion
The regular workflow is to use `git checkout` first (to switch a branch or
checkout a commit/revision of a `.dvc` file), and then run `dvc checkout` to sync
```

Reviewer: I guess this comma wasn't needed, resolving.

## Switching between versions

The regular workflow is to use `git checkout` first (to switch a branch or

> checkout a commit/revision of a `.dvc` file) and then run `dvc checkout` to sync

Reviewer: How about
```suggestion
checkout a `.dvc` file version) and then run `dvc checkout` to sync
```
? Simpler language

Author: done

<details>

### ⚙️ Expand to get the previous version of the dataset 🕑

Let's go back to the original version of the data:

```dvc
$ git checkout HEAD~1 data/data.xml.dvc
$ dvc checkout
```

Let's commit it (no need to do `dvc push` this time since this original version
of the dataset was already saved):

```dvc
$ git commit data/data.xml.dvc -m "Revert dataset updates"
```

</details>

Yes, DVC is technically not even a version control system! `.dvc` files' content

> defines data file versions. Git itself provides the version control. DVC in turn

Reviewer: 👍  How about "`.dvc` file contents define the versions of data files" ?

Author: changed

<details>

### ⚙️ Expand to get the previous version of the dataset 🕑

Let's go back to the original version of the data:

```dvc

> $ git checkout HEAD~1 data/data.xml.dvc

Reviewer: Is `HEAD~1` more common than `HEAD^`? Either one works, up to you. Just a thought

Author: `^` is an escape char in Windows so I'd avoid using it :)

Reviewer: Good thinking! Will keep it in mind.

Author: :no_entry_sign: :wind_face: :door: :door: (no-wind-doors) :P

Reviewer: I can't believe there's no window emoji.

In cases where you process very large datasets, you need an efficient mechanism
(in terms of space and performance) to share a lot of data, including different
versions. Do you use network attached storage (NAS)? Or a large external volume?
You can learn more about advanced workflows using these links:

- A shared [external cache](/doc/use-cases/shared-development-server) can be set
  up to store, version and access a lot of data on a large shared volume
  efficiently.
- A quite advanced scenario is to track and version data directly on the remote
  storage (e.g. S3). See

>   [managing external data](https://dvc.org/doc/user-guide/managing-external-data)

Reviewer: "Check out" was OK ("See" is fine too).
Capitalized title was correct though.

Author: wouldn't title case unless proper nouns (even in ref. to a doc which uses title case for its, well, title)

Reviewer: > We don't title case <abbr>s

That's a website feature (tooltips), not exactly the same as a link to a titled document. But yeah I guess we could capitalize all those tooltips to be extra correct. Let's not tho 😬 

Without capitalization here it looks like we're literally saying "See managing external data," which doesn't look like a sentence to me. "See [Any Doc Title]," seems right?

Reviewer: Or if the capitalized link gets you you we can rephrase to something like "See [this doc] on managing external data" but I wouldn't bother, we have lots of capitalized links like this everywhere.

Author: "See Proper Noun" is fine but I don't personally like awarding full proper noun status to sections of documentation. In general I'd emphasise (using e.g. italics) but - since the link already provides an underline - didn't think further emphasis was needed.

Happy to revert all links to title case if you prefer though.

It is definitely wrong to title case `<abbr>`s -- they're simply tooltip hints to dictionary definitions.

Reviewer: > Happy to revert all links to title case if you prefer

Please, I think we should revert these as long as we use the exact page titles for the links. Thanks

Author: ? I thought this was already reverted/resolved...
